backend/tests/unit/agents/test_max_iterations_failure.py

```
@@ -0,0 +1,287 @@
+        assert error.context["session_id"] == "session-123"
+        assert error.context["stage_execution_id"] == "stage-123"
+        
+        # Verify LLM was called max_iterations times
+        assert mock_llm_client.generate_response.call_count == 2
+    
+    @pytest.mark.asyncio
+    async def test_max_iterations_with_last_interaction_successful(self, controller, sample_context, mock_llm_client):
+        """Test that no exception is raised when max iterations reached but last interaction succeeded."""
+        # Mock LLM to fail first, then succeed on last attempt
+        call_count = 0
+        async def mock_generate_with_final_success(conversation, session_id, stage_execution_id=None):
```

Ruff (0.12.2)

Unused function argument: `session_id`

(ARG001)